src/Sema.zig

```diff
@@ -6076,7 +6080,31 @@ fn zirIntToEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!A
     }
 
     try sema.requireRuntimeBlock(block, src);
-    // TODO insert safety check to make sure the value matches an enum value
+    if (block.wantSafety()) {
+        const dest_int_ty = try dest_ty.enumBackingInt(sema.arena);
+        const casted_operand = try sema.intCast(block, dest_int_ty, dest_ty_src, operand, operand_src, true);
+        switch (dest_ty.tag()) {
+            .enum_nonexhaustive => {},
+            .enum_full, .enum_numbered, .enum_simple => {
+                const field_count = dest_ty.enumFieldCount();
+                if (dest_ty.tag() == .enum_simple or dest_ty.enumValues().count() == 0) {
+                    const field_count_inst = try sema.addIntUnsigned(dest_int_ty, field_count);
+                    const is_in_range = try block.addBinOp(.cmp_lt, casted_operand, field_count_inst);
+                    try sema.addSafetyCheck(block, is_in_range, .invalid_enum_value);
+                } else {
+                    // TODO: Output a switch instead of chained OR's.
```

Comment:
Hmmmm. Let's do this the right way from the start, otherwise this will harm compilation performance due to creating AIR bloat.

Suggestion: introduce a new AIR instruction called `check_enum_int` which returns true if and only if the integer is a valid tag of a given enum type. This way the backends can lower it as a switch directly, instead of having to parse the various AIR tags for a switch emitted from Sema. This would also allow the possibility for a backend to lower it to a function call in the case of enums with many tags.

stage1 already benefits from a similar arrangement because it implements runtime safety in the LLVM IR lowering pass, which is less flexible in some ways (panic function gets outputted even if it is never called) but comes with advantages, as illustrated here.

Alternative idea which you are welcome to explore: create a function in Sema called `__zig_intToEnum_E` where `E` is the fully qualified type name of the enum, which does the switch, calls panic if applicable with a custom panic message ("enum '{s}' has no tag with value '{d}'"), exactly matching the corresponding compile error. This alternative would not require introducing a new AIR tag, would not require any changes to the backends, and would pioneer the way forward to more detailed runtime safety messages.